[Spring 지하철 노선도 - 1,2단계] 마루(송채원) 미션 제출합니다.

chawani opened the pull request:
안녕하세요 마루입니다😊 리뷰 감사합니다 잘부탁드립니다~

FLYLIKEB committed:
feat: 지하철 역 등록 기능 구현

FLYLIKEB committed:
feat: 지하철 노선 생성 기능 구현

FLYLIKEB committed:
feat: 지하철 노선 조회 기능 구현

FLYLIKEB committed:
feat: 지하철 노선 수정 기능 구현

FLYLIKEB committed:
feat: 지하철 노선 제거 기능 구현

FLYLIKEB committed:
feat: StationService 조회/삭제 기능 구현

FLYLIKEB committed:
style: 코드 리포맷

FLYLIKEB committed:
build: h2 db 연결 및 스키마 설정

FLYLIKEB committed:
refactor: 지하철역 db 생성 및 데이터 연결, 스프링 빈 적용

FLYLIKEB committed:
refactor: 지하철 노선 db 생성 및 데이터 연결, 스프링 빈 적용

pkch93 commented on src/main/java/wooteco/subway/ui/StationController.java:
```diff
@@ -34,6 +45,12 @@ public ResponseEntity<List<StationResponse>> showStations() {
+    @ExceptionHandler(IllegalArgumentException.class)
+    public ResponseEntity<Void> error() {
+        return ResponseEntity.badRequest().build();
+    }
```
`IllegalArgumentException`에 대해서 LineController와 동일한 처리를 하네요.

`@ControllerAdvice`를 활용해보는건 어떠신가요?

참고: https://tecoble.techcourse.co.kr/post/2021-05-10-controller_advice_exception_handler/

pkch93 commented on src/main/java/wooteco/subway/service/StationService.java:
```diff
@@ -0,0 +1,35 @@
+            stationDao.find(station.getName());
+        } catch (IllegalArgumentException e) {
+            stationDao.save(station);
+            return stationDao.find(station.getName());
```
유효성검증을 위한 find -> 저장 -> save 결과를 반환하기 위한 find 총 3번의 쿼리가 발생하네요.

3번까지 할 필요가 있을까 싶은데요. 줄여볼수는 없을까요?

pkch93 commented on src/main/java/wooteco/subway/domain/Line.java:
```diff
@@ -0,0 +1,41 @@
+        return color;
+    }
+
+    public void update(Line line) {
```
사용하지 않는 메서드네요.

pkch93 commented on src/main/java/wooteco/subway/domain/Line.java:
```diff
@@ -0,0 +1,41 @@
+        this.color = color;
+    }
+
+    public boolean isSameName(Line line) {
```
사용하지 않는 메서드네요.

pkch93 commented on src/test/java/wooteco/subway/acceptance/LineAcceptanceTest.java:
```diff
@@ -0,0 +1,180 @@
+                .when()
+                .post("/lines")
+                .then().log().all()
+                .extract();
```
given 부분을 RestAssured로 작업하셨군요. JdbcTemplate을 주입 받아서 DB에 데이터를 넣었을수도 있었을거 같은데요. RestAssured로 API를 쏴서 준비하신 이유가 있을까요?

chawani replied:
해당 부분은 인수테스트였기 때문에 클라이언트의 입장에서 테스트를 진행하기 위하여 RestAssured를 사용하였습니다! 리뷰를 보고 given은 JdbcTemplate을 이용해도 될지 고민해봤는데, 다양한 데이터가 데이터베이스에 넣어져 있는 상태에서 테스트가 필요하다면 말씀해주신대로 하는 게 좋을 것 같습니다. 하지만 현재는 적은 데이터에 대한 테스트만 이루어지고 있어서 인수테스트라는 목적성에 맞게 포괄적으로 테스트를 하는게 좋지 않을까 생각했습니다..!
혹시 JdbcTemplate으로 테이터를 세팅하는 것과 RestAssured로 하는 것을 섞어서 쓰는 경우가 있는지 궁금합니다~

pkch93 replied:
이부분은 제 의견이니 참고만 해주셔도 좋을거 같아요.

제 기준에서 데이터 준비 과정 (given)에서 두가지를 섞어서 사용하는 경우는 없었던거 같아요.
결국 이 테스트에서 확인하고자 하는건 `GET /lines`의 결과를 테스트하는 것이니 `GET /lines`의 결과를 확인할 수 있도록 데이터를 준비하면 되지 않나 싶어요.

RestAssured나 MockMvc 같은 도구로 데이터를 준비해야한다면 이에 대한 API도 구현되어있어야하는데 그렇지 않은 경우도 종종 있었던거 같네요. 그렇다고 테스트를 위해서 굳이 API를 추가로 정의하는 것이 맞는건지는 의문이 있네요 ㅎㅎ

pkch93 replied:
RestAssured로 API를 쏴서 데이터 준비하는 과정을 별도 메서드로 빼면 어떨까요?

```java
Map<String, String> params1 = new HashMap<>();
        params1.put("name", "신분당선");
        params1.put("color", "bg-red-600");
        ExtractableResponse<Response> createResponse1 = RestAssured.given().log().all()
                .body(params1)
                .contentType(MediaType.APPLICATION_JSON_VALUE)
                .when()
                .post("/lines")
                .then().log().all()
                .extract();
```

given 절이 늘어나면서 테스트 가독성을 해친다고 생각이 드는데요. 이 부분을 별도 메서드로 빼서 가독성을 높이면 어떨까 싶습니다.

pkch93 commented on src/main/java/wooteco/subway/service/StationService.java:
```diff
@@ -0,0 +1,35 @@
+public class StationService {
+    private static final String ALREADY_IN_STATION_ERROR_MESSAGE = "이미 해당 이름의 역이 있습니다.";
+
+    private final StationDao stationDao;
```
1단계 때 빈 대신 StationDao의 static 메서드를 활용했었는데요.
static 대신 빈을 활용했을때 어떤 장단점이 있었나요 ㅎㅎ
사용하시면서 느끼신 점 말씀주시면 좋을거 같아요 :)

chawani replied:
사실 크게 장단점이 와닿지는 않았는데요,😂
생각해보면 static을 사용했을 땐 코드가 복잡해질 경우에 어떤 클래스와 연관 관계를 맺고 있는지 알아보기 어렵다는 단점이 있을 것 같습니다.
반면 bean을 사용하면 필드 선언이 필요했고 그 때문에 클래스 끼리의 연관 관계 파악이 쉬웠던 것 같습니다.
그리고 static을 사용하면서 객체지향 관점에서 벗어난다는 느낌이 있었는데, bean을 사용하니 필요한 부분에서만 클래스에 접근 할 수 있다는 점에서 코드가 객체지향적으로 바뀌었다는 생각도 들었습니다.

또 어떤 장단점이 있을까요?!

pkch93 replied:
마루가 말씀해주신거처럼 static을 무분별하게 사용하면 객체지향적보다는 유지보수가 힘든 코드로 변할수 있을거 같아요. 공유 감사해요 👍 static은 객체의 특성을 활용할 수 없기 떄문에 상속, 구현이 불가능하다는 점도 있어요.

pkch93 reviewed (changes requested):
안녕하세요 마루 👋
마루의 Spring 지하철 노선도 미션 리뷰를 맡게된 철시에요~! 반갑습니다 ㅎㅎ
잘 구현해주셨네요 ㅎㅎ 몇가지 코멘트 남겼는데 확인부탁드릴께요~!

chawani committed:
refactor: 컨트롤러 exception 처리 한번에 하도록 수정

chawani committed:
refactor: 도메인 클래스의 쓰이지 않는 메서드 삭제

chawani committed:
refactor: 객체 생성의 책임을 이동

chawani committed:
refactor: dao 및 service 수정

- dao의 queryForObject를 query를 사용하도록 수정
- dao 메서드 리네이밍
- service에서 find 호출 횟수 줄이도록 수정

chawani commented:
철시 안녕하세요~~ 마루입니다😁
리뷰 주신 것들 보고 반영해보고, 고치고 싶은 부분들도 몇 개 수정했는데 그러면서 생긴 궁금한 점들 질문드립니다:)

1. dao에서 find를 했을 때 db에 값이 없는 경우 현재처럼 dao에서 exception을 던져주는게 좋을까요? 아니면 Optional을 사용하는게 좋을까요? Optional은 신중히 사용하는게 좋다고 해서 일단 지양하는 게 좋을지 궁금하네요!
2. dao에서 db에 값이 없을 경우 던지는 예외를 NoSuchElementException, 같은 이름의 데이터를 저장하려고 할 때 던지는 예외를 IllegalStateException으로 정의했는데 적절한 예외가 맞는지 궁금합니다!
3. requestDto에는 toEntity 메서드를 정의, responseDto에는 정적 팩토리 메서드를 사용하였는데 사용하다보니 둘은 어떤 차이가 있길래 다른 방식으로 코드를 작성한거지? 하는 생각이 들더라구요. entity로 바꿔주는 메서드도 정적 메서드를 사용해도 될까요?

chawani committed:
refactor: 예외 종류 수정

pkch93 commented on src/main/java/wooteco/subway/service/StationService.java:
```diff
@@ -0,0 +1,37 @@
+
+    public Station save(Station station) {
+        try {
+            stationDao.findByName(station.getName());
```
여기서 Optional로 반환하면 어떨지를 질문주셨네요.

Optional은 값이 존재하지 않을 수 있음을 명백히 표현하기 위해서 사용될수 있도록 디자인되었는데요. 위 경우가 Optional을 활용할 수 있는 지점이지 않나 생각이 들어요.

Optional 관련한 주의사항, 가이드가 이 링크에 잘 정리가 된거 같은데요. 한번 참고해보시면 좋을거 같아요.

참고: [Java Optional 바르게 쓰기](http://homoefficio.github.io/2019/10/03/Java-Optional-%EB%B0%94%EB%A5%B4%EA%B2%8C-%EC%93%B0%EA%B8%B0/)

pkch93 commented on src/main/java/wooteco/subway/service/StationService.java:
```diff
@@ -0,0 +1,37 @@
+        } catch (NoSuchElementException exception) {
+            return stationDao.save(station);
+        }
+        throw new IllegalStateException(ALREADY_IN_STATION_ERROR_MESSAGE);
```
이전에는 `IllegalArgumentException`이었던거 같은데 `IllegalStateException`으로 변경하셨군요..!

위와 같이 변경하신 이유가 있을까요?

chawani replied:
`IllegalArgumentException`같은 경우는 잘못된 값이 인자로 넘어 왔을 때 사용하고, `IllegalStateException`은 부적절한 때에 메서드가 호출 되었음을 나타낸다고 알고 있어서 이미 같은 이름의 역이 있는 `상태`에서 부적절하게 메서드가 호출되었다고 생각해서 `IllegalStateException`을 사용하였습니다..!

pkch93 commented on src/main/java/wooteco/subway/dto/StationRequest.java:
```diff
@@ -10,6 +12,10 @@ public StationRequest(String name) {
         this.name = name;
     }
 
+    public Station toEntity() {
```
세번째 질문이 위 메서드를 static으로 정의하신다는 것이죠?
static으로 정의하면 어떤 장점이 있을까요?

추가로 toEntity 대신 Request 객체를 도메인 객체으로 변경하는 컴포넌트를 정의할수도 있는데요. 이 경우도 고려해보셨을까요?

chawani replied:
생각해보니 static으로 정의한다면 메서드에 인자로 자기자신을 넘겨줘야하는 단점이 있을 것 같습니다. 이미 해당 객체가 상태를 다 알고 있는데 굳이 인자로 넘겨 줄 필요가 없을 것 같아요..!
그리고 Request 객체를 도메인 객체로 변경하는 컴포넌트를 정의한다면 필드 변경 시 변경점이 많아진다는 단점이 있지 않을까요?? 어떻게 생각하시는지 궁금합니다!

pkch93 replied:
필드 변경시 컴포넌트의 변경 메서드 하나가 변경될거라 변경점이 그렇게 많아지는걸까 생각이 들긴하는데요 ㅎㅎ;; (클래스 기준으로는 하나 더 늘어나긴하네요)

현재는 하나의 Request 객체에서 도메인 객체로 변경하기 때문에 컴포넌트를 정의하는게 더 비용일 수는 있을거 같아요.
다만, 여러 Request 객체에서 하나의 도메인 객체로 변경을 해야하는 상황이라면 하나의 컴포넌트에서 도메인 객체 생성을 관리하는 것도 고려해볼 수 있을거 같아요. 지금 현재는 마루가 와닫지는 않을거 같은데요. 이런 점도 고려할 수 있다는 점만 참고해주시면 좋을거 같아요.

pkch93 reviewed (changes requested):
마루~! 피드백 반영 확인했습니다 :)
추가로 질문주신 부분이랑 코멘트 남겼는데요. 한번 확인부탁드릴께요~!

chawani committed:
test: 테스트 준비 과정 메서드로 추출

chawani committed:
refactor: Dao에서 Optional을 반환하도록 수정

- id로 검색하는 기능 추가

chawani committed:
test: id 조회 관련 테스트 코드 추가

pkch93 reviewed (approved):
고생하셨습니다 마루~
마루의 1, 2단계는 이만 마무리할께요 :)
다음 단계에서 뵐께요 👍
